rfcs/0000-rlp-cleanup.md

```diff
@@ -0,0 +1,331 @@
+
+### Current WIP to consider while discussing the cleaning up of the RLP
+
+1. Inheritance of the network matching rules from the upstream network object (https://github.com/Kuadrant/architecture/pull/4)
+2. Improvements to RateLimitPolicy - aka "minimal working example" without `configurations`, `limits.conditions` and `limits.variables` (https://github.com/Kuadrant/kuadrant-operator/issues/133)
+3. Implement `skip_if_absent` for the RequestHeaders action (https://github.com/Kuadrant/wasm-shim/issues/29)
+4. Rate Limiting across clusters ([doc](https://docs.google.com/document/d/1pqCODRAkNUTLB6lJfRWcv3d2Hlj9WqsGySmjrP707Vk/edit#heading=h.nzpgr3pef6uy))
+
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Given the following HTTPRoute:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1alpha2
+kind: HTTPRoute
+metadata:
+  name: toystore
+  labels:
+    app: toystore
+spec:
+  parentRefs:
+    - name: istio-ingressgateway
+      namespace: istio-system
+  hostnames: ["*.toystore.com"]
+  rules:
+    - matches:
+        - path:
+            type: PathPrefix
+            value: "/toys"
+          method: GET
+        - path:
+            type: PathPrefix
+            value: "/toys"
+          method: POST
+      backendRefs:
+        - name: toystore
+          port: 80
+```
+
+These are examples of RateLimitPolicies targeting the route:
+
+**Example 1.** Minimal example - unconditional, unqualified rate limiting
+
+```yaml
+apiVersion: kuadrant.io/v2beta1
+kind: RateLimitPolicy
+metadata:
+  name: toystore
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: toystore
+  limits:
+    - rates:
+        - limit: 5
+          unit: minute
+```
+
+**Example 2.** All you can get - conditions, counter qualifiers, multiple rates, and weights
+
+```yaml
+apiVersion: kuadrant.io/v2beta1
+kind: RateLimitPolicy
+metadata:
+  name: toystore
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: toystore
+  limits:
+    - name: writers
+      selectors:
+        # this is a condition
+        - selector: request.http.method
+          operator: eq
+          value: POST
+        # this is counter qualifier
+        - selector: metadata.auth.identity.username
+      rates:
+        - limit: 5
+          duration: 1
+          unit: minute
+        - limit: 100
+          duration: 12
+          unit: hour
+      weight: 1
+
+    - name: readers
+      selectors:
+        - selector: request.http.method
+          operator: eq
+          value: GET
+        - selector: metadata.auth.identity.username
+      rates:
+        - limit: 50
+          duration: 1
+          unit: minute
+      weight: 1
+
+    - name: read-expensive
+      selectors:
+        - selector: request.http.method
+          operator: eq
+          value: GET
+        - selector: request.http.path
```

Comment:
I like the simplicity here. When this is set, it would be good to validate that this path selector matches a rule in the HTTPRoute WDYT?

Reply:
Not sure if it's possible for everything, but in general yes. For things like headers or checking if a path prefix is contained within another, etc, such validation can be challenging.

Reply:
Yeah I think validating the path and method would be a good place to start.